salt/pillar/__init__.py

```diff
@@ -302,6 +298,119 @@ def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
             else:
                 log.error('Pillar data must be a dictionary')
 
+    @property
+    def pillar(self):
```

Comment:
``self.pillar`` is marked as a ``@property`` - that means that even though the code references ``self.pillar`` as if it's simply referencing data, the data is generated and returned by a function.  This ensures that up-to-data pillar data is available for each layer in the pillar rendering stack.  The trade-off is the generation cost for referencing ``self.pillar``.  This is why some functions do this ``pillar = self.pillar`` and then using ``pillar`` multiple times rather than multiple ``self.pillar`` references - that way it is generated once when it doesn't need to be refreshed.